src/main/scala/org/kframework/kore/interfaces/Interface.scala

```diff
@@ -0,0 +1,303 @@
+  }
+
+  trait Or extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Or {
+    def unapply(arg: Or): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Not extends Pattern {
+    def p: Pattern
+  }
+
+  object Not {
+    def unapply(arg: Not): Option[Pattern] = Some(arg.p)
+  }
+
+  trait Implies extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Implies {
+    def unapply(arg: Implies): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Exists extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object Exists {
+    def unapply(arg: Exists): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait ForAll extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object ForAll {
+    def unapply(arg: ForAll): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait Next extends Pattern {
+    def p: Pattern
+  }
+
+  object Next {
+    def unapply(arg: Next): Option[Pattern] = Some(arg.p)
+  }
+
+
+  trait Rewrite extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Rewrite {
+    def unapply(arg: Rewrite): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Equals extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Equals {
+    def unapply(arg: Equals): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait ModuleName {
+    def str: String
+  }
+
+  object ModuleName {
+    def unapply(arg: ModuleName): Option[String] = Some(arg.str)
+  }
+
+  trait Name {
+    val str: String
+  }
+
+  object Name {
+    def unapply(arg: Name): Option[String] = Some(arg.str)
+  }
+
+  trait Sort {
+    val str: String
+  }
+
+  object Sort {
+    def unapply(arg: Sort): Option[String] = Some(arg.str)
+  }
+
+  trait Symbol {
+    val str: String
+  }
+
+  object Symbol {
+    def unapply(arg: Symbol): Option[String] = Some(arg.str)
+  }
+
+  trait Value {
+    val str: String
+  }
+
+  object Value {
+    def unapply(arg: Value): Option[String] = Some(arg.str)
+  }
+
+}
+
+
+trait Builders {
+
+  import Kore._
+
+  def Definition(modules: Seq[Module], att: Attributes): Definition
+
+  def Module(name: ModuleName, sentences: Seq[Sentence], att: Attributes): Module
+
+  def Import(name: ModuleName, att: Attributes): Import
+
+  def SortDeclaration(sort: Sort, att: Attributes): SortDeclaration
+
+  def SymbolDeclaration(sort: Sort, symbol: Symbol, args: Seq[Sort], att: Attributes): SymbolDeclaration
+
+  def Rule(p: Pattern, att: Attributes): Rule
+
+  def Axiom(p: Pattern, att: Attributes): Axiom
+
+  def Attributes(att: Seq[Pattern]): Attributes
+
+  def Variable(name: Name, sort: Sort): Variable
+
+  def Application(p: Symbol, args: Seq[Pattern]): Application
+
+  def DomainValue(symbol: Symbol, value: Value): DomainValue
+
+  def Top(): Top
+
+  def Bottom(): Bottom
+
+  def And(p: Pattern, q: Pattern): And
```

Comment:
Hm, interesting.  One would expect that having the builders be more specific is better, but in this case it is the opposite.  But what's the general rule to follow here?  All builders return the least concrete type?  Why "pattern" and not "Any" then?

One could also argue the decision to apply rewrites/deduction at build time, instead of after.

Do we get any real advantage from having the return type `And` instead of `Pattern`?

In case we decide to go with `def And(p: Pattern, q: Pattern): Pattern` then we should change the KORE language back to:
```
  syntax Pattern = \and(Pattern,Pattern)
```